projects/packages/sync/src/sync-queue/class-queue-storage-table.php

```diff
@@ -0,0 +1,471 @@
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param string $queue_id The queue name this instance will be responsible for.
+	 *
+	 * @throws \Exception If queue name was not provided.
+	 */
+	public function __construct( $queue_id ) {
+		global $wpdb;
+
+		if ( empty( $queue_id ) ) {
+			// TODO what should we return here or throw an exception?
+			throw new \Exception( 'Invalid queue_id provided' );
+		}
+
+		// TODO validate the value maybe?
+		$this->queue_id = $queue_id;
+
+		// Initialize the `table_name` property with the correct prefix for easier usage in the class.
+		$this->table_name = $wpdb->prefix . $this->table_name_no_prefix;
+	}
+
+	/**
+	 * Creates the new table and updates the options to work with
+	 * the new table if it was created successfully.
+	 *
+	 * @return void
+	 */
+	public function create_table() {
+		global $wpdb;
+
+		// TODO if we run this only in the context of an upgrade, we should not include this here.
+		require_once ABSPATH . '/wp-admin/includes/upgrade.php';
+
+		$charset_collate = $wpdb->get_charset_collate();
+
+		$table_definition = "CREATE TABLE {$this->table_name} (
+			  `ID` bigint(20) NOT NULL AUTO_INCREMENT,
+			  `queue_id` varchar(50) NOT NULL,
+			  `event_id` varchar(100) NOT NULL,
+			  `event_payload` text NOT NULL,
+			  `timestamp` timestamp NOT NULL DEFAULT current_timestamp(),
+			  `object_type` varchar(100) DEFAULT NULL,
+			  `parent_object_id` bigint(20) DEFAULT NULL,
+			  PRIMARY KEY (`ID`),
+			  KEY `event_id` (`event_id`),
+			  KEY `queue_id` (`queue_id`),
+			  KEY `queue_id_event_id` (queue_id, event_id),
+			  KEY `timestamp` (`timestamp`),
+			  KEY `object_type` (`object_type`),
+			  KEY `parent_object_id` (`parent_object_id`)
+			) $charset_collate;";
+
+		/**
+		 * The function dbDelta will only return the differences. If the table exists, the result will be empty,
+		 * so let's run a check afterward to see if the table exists and is healthy.
+		 */
+		\dbDelta( $table_definition );
+	}
+
+	/**
+	 * Check if the Dedicated table actually exists.
+	 *
+	 * @return bool
+	 */
+	public function dedicated_table_exists() {
+		global $wpdb;
+
+		// Check if the table exists
+		$result = $wpdb->get_row(
+			$wpdb->prepare( 'SHOW TABLES LIKE %s', $this->table_name ),
+			ARRAY_N
+		);
+
+		if ( empty( $result ) || count( $result ) !== 1 || $result[0] !== $this->table_name ) {
+			return false;
+		}
+
+		return true;
+	}
+	/**
+	 * Check if the table is healthy, and we can read and write from/to it.
+	 *
+	 * @return bool If the dedicated table is available, and we can read and write from/to it.
+	 */
+	public function is_dedicated_table_healthy() {
+		global $wpdb;
+
+		if ( ! $this->dedicated_table_exists() ) {
+			return false;
+		}
+
+		// Try to read from the table
+
+		// Ignore the interpolated table name
+		// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+		$query = $wpdb->query( "SELECT count(`ID`) FROM {$this->table_name}" );
+
+		if ( ! $query ) {
```

Comment:
```suggestion
		if ( false === $query ) {
```

As it might return 0 results